File: scrapers/sources/diia-osvita.mjs
import { fetchHtml, politeBatch } from '../lib/fetch.mjs';

export const name = 'Дія.Освіта — курси для молоді';

const BASE = 'https://osvita.diia.gov.ua';
const TOTAL_PAGES = 21;

// Ключові слова, що вказують на релевантність для дітей/підлітків 10-18 р.
const YOUTH_KEYWORDS = [
  'молод', 'підліток', 'підлітк', 'школяр', 'учн', 'нмт', 'зно',
  'youth', 'teen', 'school', 'студент', 'абітурієнт',
  'кібербезпека для молоді', 'кібергігієна для молоді',
  'політична освіта для молоді', 'підготовка до нмт',
];

// Ключові слова, що виключають курс (для дорослих/організацій)
const ADULT_EXCLUDE = [
  'dpo', 'держслужб', 'поліц', 'ветеран', 'чизмонгер',
  'готельн', 'туристичн', 'торговельн', 'бухгалтер', 'нотаріус',
  'прокурор', 'суддя', 'адвокат', 'черліденг: тренерств',
  'тьюторинг у школі',   // для тьюторів, не учнів
  'цифрові навички для вчителів',
  'карантин: онлайн-сервіси для вчителів',
  'організація стажування молоді',
  'молодіжна робота',    // для молодіжних працівників (дорослих)
  'інклюзивна молодіжна',
  'як стати ментором',
  'наставництво: що',
  'публічна служба без',
  'covid',
  'вишивальник',
  'апарати штучної вентиляції',
];

function isYouthRelevant(title, description = '') {
  const text = `${title} ${description}`.toLowerCase();
  if (ADULT_EXCLUDE.some((kw) => text.includes(kw))) return false;
  return YOUTH_KEYWORDS.some((kw) => text.includes(kw));
}

// Парсить картки курсів зі сторінки.
// Структура: <a href="/courses/SLUG" class="...category-card...">
//   ...<h5 class="...category-card-full__title--md">TITLE</h5>
//   ...<div class="...category-card-full__description...">DESC</div>
function parseCourses(html) {
  const courses = [];
  // Витягуємо кожну <a href="/courses/..."> з її внутрішнім вмістом
  const cardRe = /<a\s+href="(\/courses\/[a-z0-9-]+)"[^>]*>([\s\S]*?)<\/a>/g;
  let m;
  while ((m = cardRe.exec(html)) !== null) {
    const path = m[1];
    const inner = m[2];
    const titleMatch = /category-card-full__title--md[^>]*>([^<]+)<\/h5>/i.exec(inner);
    if (!titleMatch) continue;
    const title = titleMatch[1].replace(/\s+/g, ' ').trim();
    const descMatch = /category-card-full__description[^>]*>([^<]*)<\/div>/i.exec(inner);
    const desc = descMatch ? descMatch[1].replace(/\s+/g, ' ').trim() : '';
    if (title.length > 3) courses.push({ path, title, desc });
  }
  return courses;
}

// Отримати опис курсу з його сторінки
async function fetchDescription(path) {
  try {
    const html = await fetchHtml(`${BASE}${path}`);
    // шукаємо мета-опис або перший параграф
    const metaDesc = /<meta[^>]+name="description"[^>]+content="([^"]{10,})"/i.exec(html);
    if (metaDesc) return metaDesc[1].trim();
    const og = /<meta[^>]+property="og:description"[^>]+content="([^"]{10,})"/i.exec(html);
    if (og) return og[1].trim();
    const p = /<p[^>]*>([^<]{30,300})<\/p>/i.exec(html);
    if (p) return p[1].replace(/\s+/g, ' ').trim();
  } catch {}
  return '';
}

export async function scrape() {
  console.log(`  Fetching ${TOTAL_PAGES} pages...`);
  const pages = Array.from({ length: TOTAL_PAGES }, (_, i) => i + 1);

  // path → { title, desc }
  const allCourses = new Map();

  await politeBatch(pages, async (page) => {
    try {
      const url = page === 1 ? `${BASE}/courses` : `${BASE}/courses?page=${page}`;
      const html = await fetchHtml(url);
      for (const { path, title, desc } of parseCourses(html)) {
        if (!allCourses.has(path)) allCourses.set(path, { title, desc });
      }
    } catch (err) {
      console.warn(`  page ${page} failed: ${err.message}`);
    }
  }, { concurrency: 3, delayMs: 500 });

  console.log(`  Found ${allCourses.size} courses total, filtering for youth...`);

  const candidates = [];
  for (const [path, { title, desc }] of allCourses) {
    if (isYouthRelevant(title, desc)) candidates.push({ path, title, desc });
  }

  const rows = [];
  for (const { path, title, desc: cardDesc } of candidates) {
    const desc = cardDesc || await fetchDescription(path);
    if (!isYouthRelevant(title, desc)) continue;

    // Визначаємо вік по вмісту
    let age_from = 14;
    let age_to = 17;
    const combined = `${title} ${desc}`.toLowerCase();
    if (/дошкільн|садок|малюк|0.{0,3}[36]/.test(combined)) { age_from = 4; age_to = 6; }
    else if (/молодш|початков|1.{0,3}4.{0,5}клас/.test(combined)) { age_from = 7; age_to = 11; }
    else if (/середн|5.{0,3}9.{0,5}клас/.test(combined)) { age_from = 10; age_to: 15; }
    else if (/нмт|зно|абітурієнт|11.{0,3}клас/.test(combined)) { age_from = 15; age_to = 17; }

    rows.push({
      title,
      summary: desc.slice(0, 500) || `Безкоштовний курс на платформі Дія.Освіта. Тема: ${title}.`,
      age_from,
      age_to,
      opportunity_type: 'course',
      categories: ['digital', 'education'],
      child_needs: [],
      format: 'Онлайн',
      cost_type: 'free',
      deadline: null,
      source_url: `${BASE}${path}`,
      source: 'Дія.Освіта',
    });
  }

  return rows;
}

File: scraper/scrapers/diia_osvita.py
"""Скрапер Дія.Освіта — безкоштовні онлайн-курси для молоді.

https://osvita.diia.gov.ua/courses — до 25 сторінок карток курсів.
Фільтр за ключовими словами; дорослі/держслужбові курси виключаються.
"""
import asyncio
import logging
import re

import httpx

logger = logging.getLogger(__name__)

SOURCE_NAME = "Дія.Освіта"
BASE_URL = "https://osvita.diia.gov.ua"
MAX_PAGES = 25

_YOUTH_KW = [
    "молод", "підліток", "підлітк", "школяр", "учн", "нмт", "зно",
    "youth", "teen", "school", "студент", "абітурієнт",
    "кібербезпека для молоді", "кібергігієна для молоді",
    "політична освіта для молоді", "підготовка до нмт",
]
_ADULT_KW = [
    "dpo", "держслужб", "поліц", "ветеран", "чизмонгер",
    "готельн", "туристичн", "торговельн", "бухгалтер", "нотаріус",
    "прокурор", "суддя", "адвокат", "черліденг: тренерств",
    "тьюторинг у школі", "цифрові навички для вчителів",
    "організація стажування молоді", "молодіжна робота",
    "інклюзивна молодіжна", "як стати ментором", "covid",
]

_CARD_RE = re.compile(r'<a\s+href="(/courses/[a-z0-9-]+)"[^>]*>([\s\S]*?)</a>', re.DOTALL)
_TITLE_RE = re.compile(r'category-card-full__title--md[^>]*>([^<]+)</h5>', re.I)
_DESC_RE = re.compile(r'category-card-full__description[^>]*>([^<]*)</div>', re.I)
_META_DESC_RE = re.compile(r'<meta[^>]+name="description"[^>]+content="([^"]{10,})"', re.I)
_OG_DESC_RE = re.compile(r'<meta[^>]+property="og:description"[^>]+content="([^"]{10,})"', re.I)

_BROWSER = {
    "User-Agent": (
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
        "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
    ),
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "uk-UA,uk;q=0.9,en;q=0.8",
    "Accept-Encoding": "gzip, deflate",
}


def _is_youth(title: str, desc: str = "") -> bool:
    text = (title + " " + desc).lower()
    if any(kw in text for kw in _ADULT_KW):
        return False
    return any(kw in text for kw in _YOUTH_KW)


def _detect_age(text: str) -> tuple[int, int]:
    t = text.lower()
    if re.search(r"дошкільн|садок|малюк", t):
        return 4, 6
    if re.search(r"молодш|початков|1.{0,4}4.{0,5}клас", t):
        return 7, 11
    if re.search(r"нмт|зно|абітурієнт|11.{0,3}клас", t):
        return 15, 17
    return 14, 17


async def fetch_all() -> list[dict]:
    async with httpx.AsyncClient(headers=_BROWSER, timeout=30.0, follow_redirects=True) as client:
        semaphore = asyncio.Semaphore(4)
        seen: dict[str, dict] = {}

        async def _fetch_listing(page: int):
            async with semaphore:
                url = f"{BASE_URL}/courses" if page == 1 else f"{BASE_URL}/courses?page={page}"
                try:
                    r = await client.get(url)
                    r.raise_for_status()
                    for m in _CARD_RE.finditer(r.text):
                        path, inner = m.group(1), m.group(2)
                        t_m = _TITLE_RE.search(inner)
                        if not t_m:
                            continue
                        title = re.sub(r"\s+", " ", t_m.group(1)).strip()
                        d_m = _DESC_RE.search(inner)
                        desc = re.sub(r"\s+", " ", d_m.group(1)).strip() if d_m else ""
                        if title and path not in seen:
                            seen[path] = {"title": title, "desc": desc}
                except Exception as e:
                    logger.warning(f"Дія.Освіта page {page}: {e}")

        await asyncio.gather(*[_fetch_listing(p) for p in range(1, MAX_PAGES + 1)])
        logger.info(f"Дія.Освіта: {len(seen)} courses total, filtering for youth")

        candidates = [(path, info) for path, info in seen.items() if _is_youth(info["title"], info["desc"])]

        async def _enrich(path: str, info: dict) -> dict | None:
            async with semaphore:
                title, desc = info["title"], info["desc"]
                if not desc:
                    try:
                        r = await client.get(f"{BASE_URL}{path}")
                        r.raise_for_status()
                        m = _META_DESC_RE.search(r.text) or _OG_DESC_RE.search(r.text)
                        desc = m.group(1).strip() if m else ""
                    except Exception:
                        pass
                if not _is_youth(title, desc):
                    return None
                age_from, age_to = _detect_age(title + " " + desc)
                return {
                    "source": SOURCE_NAME,
                    "source_url": f"{BASE_URL}{path}",
                    "raw_title": title,
                    "raw_text": (
                        f"Безкоштовний онлайн-курс на платформі Дія.Освіта.\n\n"
                        f"{desc or f'Курс: {title}. Безкоштовно. Онлайн.'}"
                    ),
                }

        results_raw = await asyncio.gather(*[_enrich(p, i) for p, i in candidates])
        results = [r for r in results_raw if r]
        logger.info(f"Дія.Освіта: {len(results)} youth-relevant courses")
        return results
